[package] PackageExporter verbose mode print package contents

lobstergrindset opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #59591 [package] skip failed __import__ statements
* **#59326 [package] PackageExporter verbose mode print package contents**

Rewrote PackageExporter to change its verbose mode `True`/`False` logic to be a leveled enum to allow for greater control over what is printed. Additionally added tracking of a node's direct predecessors to the `DiGraph` to enable easier printing of dependency informaiton.

`verbose=1` output example:
```
Package Contents:
- interned modules:
    foo__
- explicitly externed modules:
    builtins
    collections
    time
    torch
    torch._utils
    torch.nn.functional
    torch.nn.modules.activation
    torch.nn.modules.batchnorm
    torch.nn.modules.container
    torch.nn.modules.conv
    torch.optim.adam
    torch.package
- implicitly externed modules:
- mocked modules:
- pickles:
    <dcgan.model.pkl>
- denied modules which are depended upon:
    package_top
- invalid objects:
    models.DCGAN
    models.loss_criterions.ac_criterion
    models.loss_criterions.base_loss_criterions
    models.networks.DCGAN_nets
    models.utils.config
    sentencepiece
```
`verbose=2` output example:
```
Package Contents:
- interned modules:
    foo__
        * relies on:
            package_top
            sentencepiece
            time
            torch.package
- explicitly externed modules:
    builtins
        * included in:
            <dcgan.model.pkl>
    collections
        * included in:
            <dcgan.model.pkl>
    time
        * included in:
            foo__
    torch
        * included in:
            <dcgan.model.pkl>
    torch._utils
        * included in:
            <dcgan.model.pkl>
    torch.nn.functional
        * included in:
            <dcgan.model.pkl>
    torch.nn.modules.activation
        * included in:
            <dcgan.model.pkl>
    torch.nn.modules.batchnorm
        * included in:
            <dcgan.model.pkl>
    torch.nn.modules.container
        * included in:
            <dcgan.model.pkl>
    torch.nn.modules.conv
        * included in:
            <dcgan.model.pkl>
    torch.optim.adam
        * included in:
            <dcgan.model.pkl>
    torch.package
        * included in:
            foo__
- implicitly externed modules:
- mocked modules:
- pickles:
    <dcgan.model.pkl>
        * relies on:
            builtins
            collections
            models.DCGAN
            models.loss_criterions.ac_criterion
            models.loss_criterions.base_loss_criterions
            models.networks.DCGAN_nets
            models.utils.config
            torch
            torch._utils
            torch.nn.functional
            torch.nn.modules.activation
            torch.nn.modules.batchnorm
            torch.nn.modules.container
            torch.nn.modules.conv
            torch.optim.adam
- denied modules which are depended upon:
    package_top
        * included in:
            foo__
- invalid objects:
    models.DCGAN
        * included in:
            <dcgan.model.pkl>
    models.loss_criterions.ac_criterion
        * included in:
            <dcgan.model.pkl>
    models.loss_criterions.base_loss_criterions
        * included in:
            <dcgan.model.pkl>
    models.networks.DCGAN_nets
        * included in:
            <dcgan.model.pkl>
    models.utils.config
        * included in:
            <dcgan.model.pkl>
    sentencepiece
        * included in:
            foo__
Dependency graph for exported package: 
https://dreampuf.github.io/GraphvizOnline/#digraph%20G%20%7B%0Arankdir%20%3D%20LR%3B%0Anode%20%5Bshape%3Dbox%5D%3B%0A%0A%22%3Cdcgan.model.pkl%3E%22%20-%3E%20%22models.DCGAN%22%3B%0A%22%3Cdcgan.model.pkl%3E%22%20-%3E%20%22models.utils.config%22%3B%0A%22%3Cdcgan.model.pkl%3E%22%20-%3E%20%22torch%22%3B%0A%22%3Cdcgan.model.pkl%3E%22%20-%3E%20%22models.loss_criterions.ac_criterion%22%3B%0A%22%3Cdcgan.model.pkl%3E%22%20-%3E%20%22torch._utils%22%3B%0A%22%3Cdcgan.model.pkl%3E%22%20-%3E%20%22collections%22%3B%0A%22%3Cdcgan.model.pkl%3E%22%20-%3E%20%22models.loss_criterions.base_loss_criterions%22%3B%0A%22%3Cdcgan.model.pkl%3E%22%20-%3E%20%22torch.nn.functional%22%3B%0A%22%3Cdcgan.model.pkl%3E%22%20-%3E%20%22models.networks.DCGAN_nets%22%3B%0A%22%3Cdcgan.model.pkl%3E%22%20-%3E%20%22builtins%22%3B%0A%22%3Cdcgan.model.pkl%3E%22%20-%3E%20%22torch.nn.modules.container%22%3B%0A%22%3Cdcgan.model.pkl%3E%22%20-%3E%20%22torch.nn.modules.conv%22%3B%0A%22%3Cdcgan.model.pkl%3E%22%20-%3E%20%22torch.nn.modules.activation%22%3B%0A%22%3Cdcgan.model.pkl%3E%22%20-%3E%20%22torch.nn.modules.batchnorm%22%3B%0A%22%3Cdcgan.model.pkl%3E%22%20-%3E%20%22torch.optim.adam%22%3B%0A%22foo__%22%20-%3E%20%22torch.package%22%3B%0A%22foo__%22%20-%3E%20%22time%22%3B%0A%22foo__%22%20-%3E%20%22sentencepiece%22%3B%0A%22foo__%22%20-%3E%20%22package_top%22%3B%0A%7D%0A
```

Differential Revision: [D28871216](https://our.internmc.facebook.com/intern/diff/D28871216)

Commit:
[package] PackageExporter verbose mode print package contents

[ghstack-poisoned]

Commit:
Update on "[package] PackageExporter verbose mode print package contents"


Added additional logic to PackageExporter to print out package's contents organized by dependency node type.

Example output:
```
Package Contents:
- interned modules:
    foo__
- explicitly externed modules:
- implicitly externed modules:
    builtins
    collections
    time
    torch
    torch._utils
    torch.nn.functional
    torch.nn.modules.activation
    torch.nn.modules.batchnorm
    torch.nn.modules.container
    torch.nn.modules.conv
    torch.optim.adam
    torch.package
- mocked modules:
- pickles:
    <dcgan.model.pkl>
- invalid objects:
    models.DCGAN
    models.loss_criterions.ac_criterion
    models.loss_criterions.base_loss_criterions
    models.networks.DCGAN_nets
    models.utils.config
    sentencepiece
```

[ghstack-poisoned]

Commit:
Update on "[package] PackageExporter verbose mode print package contents"


Added additional logic to PackageExporter to print out package's contents organized by dependency node type.

Example output:
```
Package Contents:
- interned modules:
    foo__
- explicitly externed modules:
- implicitly externed modules:
    builtins
    collections
    time
    torch
    torch._utils
    torch.nn.functional
    torch.nn.modules.activation
    torch.nn.modules.batchnorm
    torch.nn.modules.container
    torch.nn.modules.conv
    torch.optim.adam
    torch.package
- mocked modules:
- pickles:
    <dcgan.model.pkl>
- invalid objects:
    models.DCGAN
    models.loss_criterions.ac_criterion
    models.loss_criterions.base_loss_criterions
    models.networks.DCGAN_nets
    models.utils.config
    sentencepiece
```

[ghstack-poisoned]

Lilyjjo commented:
@Lilyjjo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28871216).

SplitInfinity commented on torch/package/package_exporter.py:
```diff
@@ -830,6 +823,67 @@ def _execute_dependency_graph(self):
+                else:
+                    interned_mods.append(module_name)
+            else:
+                invalid_nodes.append(module_name)
```
Also, invalid here could mean that no action was taken, or that the module matched a deny pattern but is depended on by some module. Should we distinguish between the two cases?

SplitInfinity reviewed (commented):
I noticed that some of the verbose print messages that you moved were in other functions that are called before `close`. Is there a chance that an error can occur before `close` is called (meaning that these messages no longer have the chance to appear). 

I think the answer to that question is "no" because of the code reorg that was done during the PR that introduced `intern`, but I just want to double check.

SplitInfinity commented on torch/package/_digraph.py:
```diff
@@ -11,6 +11,7 @@ def __init__(self):
         # Nested dict of node -> successor node -> nothing.
         # (didn't implement edge data)
         self._succ = {}
+        self._pred = {}
```
Can you add a comment for this attribute?

SplitInfinity commented on torch/package/package_exporter.py:
```diff
@@ -382,11 +383,7 @@ def require_module_if_not_provided(self, module_name: str, dependencies=True):
-                    f"implicitly adding {module_name} to external modules "
-                    f"since it is part of the standard library and is a dependency."
-                )
+            self._implicitly_externed.append(module_name)
```
Since `DiGraph` allows you to add arbitrary node attributes, consider record the fact that an extern is implicit in the graph node itself and so you can avoid introducing this new attribute.

Lilyjjo commented on torch/package/package_exporter.py:
```diff
@@ -171,6 +171,7 @@ def __init__(
         f: Union[str, Path, BinaryIO],
         importer: Union[Importer, Sequence[Importer]] = sys_importer,
         verbose: bool = True,
+        verbosity_lvl: int = 0,
```
Yeah, I would've preferred that as an impl, I just was cautious to change the base API. The conversion of True/False to 0/1 is automatic, so it should be fine

suo commented:
clearing my queue for now, please re-add me when other comments are addressed—I want to review any changes to the dependency graph we're making.

Commit:
Update on "[package] PackageExporter verbose mode print package contents"


Rewrote PackageExporter verbose mode logic and added verbosity level toggle to give users more control over amount of information printed. 

`verbosity_lvl = 0` output example:
```
Package Contents:
- interned modules:
    foo__
- explicitly externed modules:
- implicitly externed modules:
    builtins
    collections
    time
    torch
    torch._utils
    torch.nn.functional
    torch.nn.modules.activation
    torch.nn.modules.batchnorm
    torch.nn.modules.container
    torch.nn.modules.conv
    torch.optim.adam
    torch.package
- mocked modules:
- pickles:
    <dcgan.model.pkl>
- invalid objects:
    models.DCGAN
    models.loss_criterions.ac_criterion
    models.loss_criterions.base_loss_criterions
    models.networks.DCGAN_nets
    models.utils.config
    package_top
    sentencepiece
```
`verbosity_lvl = 1` output example:
```
Package Contents:
- interned modules:
    foo__
        * relies on:
            package_top
            sentencepiece
            time
            torch.package
- explicitly externed modules:
- implicitly externed modules:
    builtins
        * included in:
            <dcgan.model.pkl>
    collections
        * included in:
            <dcgan.model.pkl>
    time
        * included in:
            foo__
    torch
        * included in:
            <dcgan.model.pkl>
    torch._utils
        * included in:
            <dcgan.model.pkl>
    torch.nn.functional
        * included in:
            <dcgan.model.pkl>
    torch.nn.modules.activation
        * included in:
            <dcgan.model.pkl>
    torch.nn.modules.batchnorm
        * included in:
            <dcgan.model.pkl>
    torch.nn.modules.container
        * included in:
            <dcgan.model.pkl>
    torch.nn.modules.conv
        * included in:
            <dcgan.model.pkl>
    torch.optim.adam
        * included in:
            <dcgan.model.pkl>
    torch.package
        * included in:
            foo__
- mocked modules:
- pickles:
    <dcgan.model.pkl>
        * relies on:
            builtins
            collections
            models.DCGAN
            models.loss_criterions.ac_criterion
            models.loss_criterions.base_loss_criterions
            models.networks.DCGAN_nets
            models.utils.config
            torch
            torch._utils
            torch.nn.functional
            torch.nn.modules.activation
            torch.nn.modules.batchnorm
            torch.nn.modules.container
            torch.nn.modules.conv
            torch.optim.adam
- invalid objects:
    models.DCGAN
        * included in:
            <dcgan.model.pkl>
    models.loss_criterions.ac_criterion
        * included in:
            <dcgan.model.pkl>
    models.loss_criterions.base_loss_criterions
        * included in:
            <dcgan.model.pkl>
    models.networks.DCGAN_nets
        * included in:
            <dcgan.model.pkl>
    models.utils.config
        * included in:
            <dcgan.model.pkl>
    package_top
        * included in:
            foo__
    sentencepiece
        * included in:
            foo__
Dependency graph for exported package: 
https://dreampuf.github.io/GraphvizOnline/#digraph%20G%20%7B%0Arankdir%20%3D%20LR%3B%0Anode%20%5Bshape%3Dbox%5D%3B%0A%0A%22%3Cdcgan.model.pkl%3E%22%20-%3E%20%22models.DCGAN%22%3B%0A%22%3Cdcgan.model.pkl%3E%22%20-%3E%20%22models.utils.config%22%3B%0A%22%3Cdcgan.model.pkl%3E%22%20-%3E%20%22torch%22%3B%0A%22%3Cdcgan.model.pkl%3E%22%20-%3E%20%22models.loss_criterions.ac_criterion%22%3B%0A%22%3Cdcgan.model.pkl%3E%22%20-%3E%20%22torch._utils%22%3B%0A%22%3Cdcgan.model.pkl%3E%22%20-%3E%20%22collections%22%3B%0A%22%3Cdcgan.model.pkl%3E%22%20-%3E%20%22models.loss_criterions.base_loss_criterions%22%3B%0A%22%3Cdcgan.model.pkl%3E%22%20-%3E%20%22torch.nn.functional%22%3B%0A%22%3Cdcgan.model.pkl%3E%22%20-%3E%20%22models.networks.DCGAN_nets%22%3B%0A%22%3Cdcgan.model.pkl%3E%22%20-%3E%20%22builtins%22%3B%0A%22%3Cdcgan.model.pkl%3E%22%20-%3E%20%22torch.nn.modules.container%22%3B%0A%22%3Cdcgan.model.pkl%3E%22%20-%3E%20%22torch.nn.modules.conv%22%3B%0A%22%3Cdcgan.model.pkl%3E%22%20-%3E%20%22torch.nn.modules.activation%22%3B%0A%22%3Cdcgan.model.pkl%3E%22%20-%3E%20%22torch.nn.modules.batchnorm%22%3B%0A%22%3Cdcgan.model.pkl%3E%22%20-%3E%20%22torch.optim.adam%22%3B%0A%22foo__%22%20-%3E%20%22torch.package%22%3B%0A%22foo__%22%20-%3E%20%22time%22%3B%0A%22foo__%22%20-%3E%20%22sentencepiece%22%3B%0A%22foo__%22%20-%3E%20%22package_top%22%3B%0A%7D%0A
```

Differential Revision: [D28871216](https://our.internmc.facebook.com/intern/diff/D28871216)

[ghstack-poisoned]

Lilyjjo commented:
@Lilyjjo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28871216).

Lilyjjo commented:
@SplitInfinity I took all of your suggestions! Thanks for them. I'm still unsure about if changing the `PackageExporter`'s `verbose` flag from a bool to an int is a good idea. I'm also unsure of how to best document the verbosity enum categories if we go that route

SplitInfinity commented on torch/package/package_exporter.py:
```diff
@@ -830,6 +833,81 @@ def _execute_dependency_graph(self):
+            tabs += tab
+            for item in sorted(entries):
+                print(f"{tabs}{item}")
+                if recurse and self.verbose > 1:
```
```suggestion
                if recurse and self.verbose > VerbosityLevel.PACKAGE_INFO:
```

SplitInfinity commented on torch/package/package_exporter.py:
```diff
@@ -830,6 +833,81 @@ def _execute_dependency_graph(self):
+        print_pretty("pickles", pickled_objs)
+        print_pretty("denied modules which are depended upon", denied_nodes)
+        print_pretty("invalid objects", invalid_nodes)
+        if self.verbose > 1:
```
```suggestion
        if self.verbose > VerbosityLevel.PACKAGE_INFO:
```

Commit:
Update on "[package] PackageExporter verbose mode print package contents"


Rewrote PackageExporter to change its verbose mode `True`/`False` logic to be a leveled enum to allow for greater control over what is printed. Additionally added tracking of a node's direct predecessors to the `DiGraph` to enable easier printing of dependency informaiton.

`verbose=1` output example:
```
Package Contents:
- interned modules:
    foo__
- explicitly externed modules:
    builtins
    collections
    time
    torch
    torch._utils
    torch.nn.functional
    torch.nn.modules.activation
    torch.nn.modules.batchnorm
    torch.nn.modules.container
    torch.nn.modules.conv
    torch.optim.adam
    torch.package
- implicitly externed modules:
- mocked modules:
- pickles:
    <dcgan.model.pkl>
- denied modules which are depended upon:
    package_top
- invalid objects:
    models.DCGAN
    models.loss_criterions.ac_criterion
    models.loss_criterions.base_loss_criterions
    models.networks.DCGAN_nets
    models.utils.config
    sentencepiece
```
`verbose=2` output example:
```
Package Contents:
- interned modules:
    foo__
        * relies on:
            package_top
            sentencepiece
            time
            torch.package
- explicitly externed modules:
    builtins
        * included in:
            <dcgan.model.pkl>
    collections
        * included in:
            <dcgan.model.pkl>
    time
        * included in:
            foo__
    torch
        * included in:
            <dcgan.model.pkl>
    torch._utils
        * included in:
            <dcgan.model.pkl>
    torch.nn.functional
        * included in:
            <dcgan.model.pkl>
    torch.nn.modules.activation
        * included in:
            <dcgan.model.pkl>
    torch.nn.modules.batchnorm
        * included in:
            <dcgan.model.pkl>
    torch.nn.modules.container
        * included in:
            <dcgan.model.pkl>
    torch.nn.modules.conv
        * included in:
            <dcgan.model.pkl>
    torch.optim.adam
        * included in:
            <dcgan.model.pkl>
    torch.package
        * included in:
            foo__
- implicitly externed modules:
- mocked modules:
- pickles:
    <dcgan.model.pkl>
        * relies on:
            builtins
            collections
            models.DCGAN
            models.loss_criterions.ac_criterion
            models.loss_criterions.base_loss_criterions
            models.networks.DCGAN_nets
            models.utils.config
            torch
            torch._utils
            torch.nn.functional
            torch.nn.modules.activation
            torch.nn.modules.batchnorm
            torch.nn.modules.container
            torch.nn.modules.conv
            torch.optim.adam
- denied modules which are depended upon:
    package_top
        * included in:
            foo__
- invalid objects:
    models.DCGAN
        * included in:
            <dcgan.model.pkl>
    models.loss_criterions.ac_criterion
        * included in:
            <dcgan.model.pkl>
    models.loss_criterions.base_loss_criterions
        * included in:
            <dcgan.model.pkl>
    models.networks.DCGAN_nets
        * included in:
            <dcgan.model.pkl>
    models.utils.config
        * included in:
            <dcgan.model.pkl>
    sentencepiece
        * included in:
            foo__
Dependency graph for exported package: 
https://dreampuf.github.io/GraphvizOnline/#digraph%20G%20%7B%0Arankdir%20%3D%20LR%3B%0Anode%20%5Bshape%3Dbox%5D%3B%0A%0A%22%3Cdcgan.model.pkl%3E%22%20-%3E%20%22models.DCGAN%22%3B%0A%22%3Cdcgan.model.pkl%3E%22%20-%3E%20%22models.utils.config%22%3B%0A%22%3Cdcgan.model.pkl%3E%22%20-%3E%20%22torch%22%3B%0A%22%3Cdcgan.model.pkl%3E%22%20-%3E%20%22models.loss_criterions.ac_criterion%22%3B%0A%22%3Cdcgan.model.pkl%3E%22%20-%3E%20%22torch._utils%22%3B%0A%22%3Cdcgan.model.pkl%3E%22%20-%3E%20%22collections%22%3B%0A%22%3Cdcgan.model.pkl%3E%22%20-%3E%20%22models.loss_criterions.base_loss_criterions%22%3B%0A%22%3Cdcgan.model.pkl%3E%22%20-%3E%20%22torch.nn.functional%22%3B%0A%22%3Cdcgan.model.pkl%3E%22%20-%3E%20%22models.networks.DCGAN_nets%22%3B%0A%22%3Cdcgan.model.pkl%3E%22%20-%3E%20%22builtins%22%3B%0A%22%3Cdcgan.model.pkl%3E%22%20-%3E%20%22torch.nn.modules.container%22%3B%0A%22%3Cdcgan.model.pkl%3E%22%20-%3E%20%22torch.nn.modules.conv%22%3B%0A%22%3Cdcgan.model.pkl%3E%22%20-%3E%20%22torch.nn.modules.activation%22%3B%0A%22%3Cdcgan.model.pkl%3E%22%20-%3E%20%22torch.nn.modules.batchnorm%22%3B%0A%22%3Cdcgan.model.pkl%3E%22%20-%3E%20%22torch.optim.adam%22%3B%0A%22foo__%22%20-%3E%20%22torch.package%22%3B%0A%22foo__%22%20-%3E%20%22time%22%3B%0A%22foo__%22%20-%3E%20%22sentencepiece%22%3B%0A%22foo__%22%20-%3E%20%22package_top%22%3B%0A%7D%0A
```

Differential Revision: [D28871216](https://our.internmc.facebook.com/intern/diff/D28871216)

[ghstack-poisoned]

Lilyjjo commented:
@Lilyjjo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28871216).

Commit:
Update on "[package] PackageExporter verbose mode print package contents"


Rewrote PackageExporter to change its verbose mode `True`/`False` logic to be a leveled enum to allow for greater control over what is printed. Additionally added tracking of a node's direct predecessors to the `DiGraph` to enable easier printing of dependency informaiton.

`verbose=1` output example:
```
Package Contents:
- interned modules:
    foo__
- explicitly externed modules:
    builtins
    collections
    time
    torch
    torch._utils
    torch.nn.functional
    torch.nn.modules.activation
    torch.nn.modules.batchnorm
    torch.nn.modules.container
    torch.nn.modules.conv
    torch.optim.adam
    torch.package
- implicitly externed modules:
- mocked modules:
- pickles:
    <dcgan.model.pkl>
- denied modules which are depended upon:
    package_top
- invalid objects:
    models.DCGAN
    models.loss_criterions.ac_criterion
    models.loss_criterions.base_loss_criterions
    models.networks.DCGAN_nets
    models.utils.config
    sentencepiece
```
`verbose=2` output example:
```
Package Contents:
- interned modules:
    foo__
        * relies on:
            package_top
            sentencepiece
            time
            torch.package
- explicitly externed modules:
    builtins
        * included in:
            <dcgan.model.pkl>
    collections
        * included in:
            <dcgan.model.pkl>
    time
        * included in:
            foo__
    torch
        * included in:
            <dcgan.model.pkl>
    torch._utils
        * included in:
            <dcgan.model.pkl>
    torch.nn.functional
        * included in:
            <dcgan.model.pkl>
    torch.nn.modules.activation
        * included in:
            <dcgan.model.pkl>
    torch.nn.modules.batchnorm
        * included in:
            <dcgan.model.pkl>
    torch.nn.modules.container
        * included in:
            <dcgan.model.pkl>
    torch.nn.modules.conv
        * included in:
            <dcgan.model.pkl>
    torch.optim.adam
        * included in:
            <dcgan.model.pkl>
    torch.package
        * included in:
            foo__
- implicitly externed modules:
- mocked modules:
- pickles:
    <dcgan.model.pkl>
        * relies on:
            builtins
            collections
            models.DCGAN
            models.loss_criterions.ac_criterion
            models.loss_criterions.base_loss_criterions
            models.networks.DCGAN_nets
            models.utils.config
            torch
            torch._utils
            torch.nn.functional
            torch.nn.modules.activation
            torch.nn.modules.batchnorm
            torch.nn.modules.container
            torch.nn.modules.conv
            torch.optim.adam
- denied modules which are depended upon:
    package_top
        * included in:
            foo__
- invalid objects:
    models.DCGAN
        * included in:
            <dcgan.model.pkl>
    models.loss_criterions.ac_criterion
        * included in:
            <dcgan.model.pkl>
    models.loss_criterions.base_loss_criterions
        * included in:
            <dcgan.model.pkl>
    models.networks.DCGAN_nets
        * included in:
            <dcgan.model.pkl>
    models.utils.config
        * included in:
            <dcgan.model.pkl>
    sentencepiece
        * included in:
            foo__
Dependency graph for exported package: 
https://dreampuf.github.io/GraphvizOnline/#digraph%20G%20%7B%0Arankdir%20%3D%20LR%3B%0Anode%20%5Bshape%3Dbox%5D%3B%0A%0A%22%3Cdcgan.model.pkl%3E%22%20-%3E%20%22models.DCGAN%22%3B%0A%22%3Cdcgan.model.pkl%3E%22%20-%3E%20%22models.utils.config%22%3B%0A%22%3Cdcgan.model.pkl%3E%22%20-%3E%20%22torch%22%3B%0A%22%3Cdcgan.model.pkl%3E%22%20-%3E%20%22models.loss_criterions.ac_criterion%22%3B%0A%22%3Cdcgan.model.pkl%3E%22%20-%3E%20%22torch._utils%22%3B%0A%22%3Cdcgan.model.pkl%3E%22%20-%3E%20%22collections%22%3B%0A%22%3Cdcgan.model.pkl%3E%22%20-%3E%20%22models.loss_criterions.base_loss_criterions%22%3B%0A%22%3Cdcgan.model.pkl%3E%22%20-%3E%20%22torch.nn.functional%22%3B%0A%22%3Cdcgan.model.pkl%3E%22%20-%3E%20%22models.networks.DCGAN_nets%22%3B%0A%22%3Cdcgan.model.pkl%3E%22%20-%3E%20%22builtins%22%3B%0A%22%3Cdcgan.model.pkl%3E%22%20-%3E%20%22torch.nn.modules.container%22%3B%0A%22%3Cdcgan.model.pkl%3E%22%20-%3E%20%22torch.nn.modules.conv%22%3B%0A%22%3Cdcgan.model.pkl%3E%22%20-%3E%20%22torch.nn.modules.activation%22%3B%0A%22%3Cdcgan.model.pkl%3E%22%20-%3E%20%22torch.nn.modules.batchnorm%22%3B%0A%22%3Cdcgan.model.pkl%3E%22%20-%3E%20%22torch.optim.adam%22%3B%0A%22foo__%22%20-%3E%20%22torch.package%22%3B%0A%22foo__%22%20-%3E%20%22time%22%3B%0A%22foo__%22%20-%3E%20%22sentencepiece%22%3B%0A%22foo__%22%20-%3E%20%22package_top%22%3B%0A%7D%0A
```

Differential Revision: [D28871216](https://our.internmc.facebook.com/intern/diff/D28871216)

[ghstack-poisoned]

Lilyjjo commented:
@Lilyjjo has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28871216).

suo commented on torch/package/package_exporter.py:
```diff
@@ -852,15 +853,89 @@ def _execute_dependency_graph(self):
+        invalid_nodes = []
+        denied_nodes = []
+
+        for module_name, attrs in self.dependency_graph.nodes.items():
```
I don't think we should hard code these categories here. Maybe each node should get an additional attribute with what category it should be printed under and we just group nodes by that attribute.

suo commented on torch/package/package_exporter.py:
```diff
@@ -170,7 +178,7 @@ def __init__(
         f: Union[str, Path, BinaryIO],
         importer: Union[Importer, Sequence[Importer]] = sys_importer,
-        verbose: bool = True,
+        verbose: int = VerbosityLevel.NONE,
```
This is a BC breaking change. We don't commit to BC yet for this API, but it will break internal callers, right? Or is there implicit casting going on.

SplitInfinity replied:
We can go fix all of the uses, right? 

We could also keep `verbose` around just for a little bit to provide BC and then remove it later?

Lilyjjo replied:
True/False will cast to the ints implicitly, it'll just break all the static typing checks. I think fixing them wouldn't be too difficult

suo commented on torch/package/package_exporter.py:
```diff
@@ -852,15 +853,89 @@ def _execute_dependency_graph(self):
+        print_pretty("pickles", pickled_objs)
+        print_pretty("denied modules which are depended upon", denied_nodes)
+        print_pretty("invalid objects", invalid_nodes)
+        if self.verbose > VerbosityLevel.PACKAGE_INFO:
```
There's some duplication in intent between this code and the error messaging code. Both will print listings of modules grouped by their characteristics (broken/denied/etc.). We should decide what information we want to go into the verbose logs vs. the error message.

suo commented on torch/package/package_exporter.py:
```diff
@@ -180,8 +188,11 @@ def __init__(
+            verbose: Print information about dependency resolution to stdout. Higher values will increase
+                amount of information printed. ``VerbosityLevel.NONE=0`` will print out nothing,
+                ``VerbosityLevel.PACKAGE_INFO=1`` will print out the package contents sorted by category, and
+                ``VerbosityLevel.PACKAGE_AND_DEPENDENCY_INFO=3`` will print out the package contents with dependency
```
=2, not 3

suo commented:
Overall: I'm not super clear on the purpose of verbose mode. Like, what info would the user be looking for and is the best way to present that information showing in stdout?

The most obvious use case to me is for debugging when something goes wrong with a package. But that is already covered by `PackagingError` and it's not clear how verbose mode is different.

Lilyjjo commented:
The extra functionality of the verbose mode is printing out the dependencies and showing in totality what is in a package. That functionality could also be better put into explicit calls instead of a blanket print out statement ...
